pkg/server/server.go

```diff
@@ -4367,6 +4363,158 @@ func (s *BgpServer) ResetRpki(ctx context.Context, r *api.ResetRpkiRequest) erro
 	}, false)
 }
 
+func toPathApiUtil(path *table.Path) *apiutil.Path {
+	// Best and SendMaxFiltered are set in ListPath API
+	p := &apiutil.Path{
+		Nlri:       path.GetNlri(),
+		Age:        path.GetTimestamp().Unix(),
+		Attrs:      path.GetPathAttrs(),
+		Stale:      path.IsStale(),
+		Withdrawal: path.IsWithdraw,
+	}
+	if path.GetSource() != nil {
+		p.SourceASN = path.GetSource().AS
+		p.SourceID = path.GetSource().ID
+		p.NeighborIP = path.GetSource().Address
+	}
+	return p
+}
+
+type WatchEventMessageCallbacks struct {
+	OnPathUpdate func([]*apiutil.Path, time.Time)
+	OnBestPath   func([]*apiutil.Path, time.Time)
+	OnPathEor    func([]*apiutil.Path, time.Time)
+	OnPeerUpdate func(*apiutil.WatchEventMessage_PeerEvent, time.Time)
+}
+
+func (s *BgpServer) WatchEventMessages(ctx context.Context, r *api.WatchEventRequest, callbacks WatchEventMessageCallbacks) error {
+	if r == nil {
+		return fmt.Errorf("nil request")
+	}
+
+	opts := make([]watchOption, 0)
+	if r.GetPeer() != nil {
+		opts = append(opts, watchPeer())
+	}
+	if t := r.GetTable(); t != nil {
+		for _, filter := range t.Filters {
+			switch filter.Type {
+			case api.WatchEventRequest_Table_Filter_TYPE_BEST:
+				opts = append(opts, watchBestPath(filter.Init))
+			case api.WatchEventRequest_Table_Filter_TYPE_ADJIN:
+				opts = append(opts, watchUpdate(filter.Init, filter.PeerAddress, filter.PeerGroup))
+			case api.WatchEventRequest_Table_Filter_TYPE_POST_POLICY:
+				opts = append(opts, watchPostUpdate(filter.Init, filter.PeerAddress, filter.PeerGroup))
+			case api.WatchEventRequest_Table_Filter_TYPE_EOR:
+				opts = append(opts, watchEor(filter.Init))
+			default:
+				return status.Errorf(codes.InvalidArgument, "unknown filter type %s", filter.Type)
+			}
+		}
+	}
+	if len(opts) == 0 {
+		return fmt.Errorf("no events to watch")
+	}
+	w := s.watch(opts...)
+
+	go func() {
```

Comment:
I think that we have the similar code for gRPC API. Is there a way to share something, like by passing a closure that handles data conversion as an argument?